pkgs/applications/science/math/singular/default.nix

```diff
@@ -1,57 +1,96 @@
-{ stdenv, fetchurl, gmp, bison, perl, ncurses, readline, coreutils, pkg-config
+{ stdenv, fetchFromGitHub, gmp, bison, perl, ncurses, readline, coreutils, pkg-config
 , lib
 , fetchpatch
 , autoreconfHook
+, sharutils
 , file
 , flint
 , ntl
 , cddlib
-, enableFactory ? true
+, gfan
+, lrcalc
+, doxygen
+, graphviz
+# upstream generates docs with texinfo 4. later versions of texinfo
+# use letters instead of numbers for post-appendix chapters, and we
+# want it to match the upstream format because sage depends on it.
+, texinfo4
+, texlive
+, enableDocs ? true
 , enableGfanlib ? true
 }:
 
 stdenv.mkDerivation rec {
   pname = "singular";
-  version = "4.1.1p2";
-
-  src = let
-    # singular sorts its tarballs in directories by base release (without patch version)
-    # for example 4.1.1p1 will be in the directory 4-1-1
-    baseVersion = builtins.head (lib.splitString "p" version);
-    urlVersion = builtins.replaceStrings [ "." ] [ "-" ] baseVersion;
-  in
-  fetchurl {
-    url = "http://www.mathematik.uni-kl.de/ftp/pub/Math/Singular/SOURCES/${urlVersion}/singular-${version}.tar.gz";
-    sha256 = "07x9kri8vl4galik7lr6pscq3c51n8570pyw64i7gbj0m706f7wf";
+  version = "4.2.0p2";
+
+  # since the tarball does not contain tests or documentation (and
+  # there is no separate tests tarball for 4.2.0), we fetch from
+  # GitHub.
+  src = fetchFromGitHub {
+    owner = "Singular";
+    repo = "Singular";
+
+    # 4.2.0p2 is not tagged, but the tarball matches commit
+    # 6f68939ddf612d96e3caaaaa8275f77613ac1da8. the commit below has
+    # two extra fixes.
+    rev = "3cda50c00a849455efa2502e56596955491a353a";
+    sha256 = "sha256-OizPhGE6L2LTOrKfeDdDB6BSdvYkDVXvbbYjV14hnHM=";
+
+    # if a release is tagged it will be in the format below.
+    # rev = "Release${lib.replaceStrings ["."] ["-"] version}";
+
+    # the repository's .gitattributes file contains the lines "/Tst/
+    # export-ignore" and "/doc/ export-ignore" so some directories are
+    # not included in the tarball downloaded by fetchzip. setting
+    # fetchSubmodules works around this by using fetchgit instead of
+    # fetchzip.
+    fetchSubmodules = true;
   };
 
+  patches = [
```

Comment:
All these patches and the `sed` line in `postPatch` are not nixos specific right? Are there corresponding upstream issues?

Reply:
Yes, agreed. Here are the things I need to do:

- [ ] Ask Singular developers if aarch64 support is a goal. If so, investigate the vspace-related hang.
- [ ] File an upstream bug for docbuilding the manual without having all optional dependencies (I think it's supposed to work).
- [ ] Submit a patch replacing the weirdly-encoded character by `@cdot{}`.

I will tick those boxes as I do those items.

Reply:
Actually @timokau, if you do have an aarch64 machine available, running Singular tests (removing the `disable-vspace.patch` patch) under gdb to diagnose vspace-related problems would be great. I don't think upstream will be able to solve the bug with the information I have.

Reply:
Going above and beyond, as always :)

I do have access to [the community build box](https://github.com/nix-community/aarch64-build-box). My time for tinkering on open source things is unfortunately limited recently (it's always a question of priorities of course, but other things rank higher for me right now). I could run some tests for you, but digging into the hanging issue would probably require quite some time in the debugger.

Maybe upstream would already appreciate a report of the issue.  They might consider applying the "disable vspace" workaround, even if they don't want to invest much time into aarch64 support.